// @fuel-ts/program/dist/index.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/types.ts
var AbstractProgram = class {
  static {
    __name(this, "AbstractProgram");
  }
};
var AbstractContract = class extends AbstractProgram {
  static {
    __name(this, "AbstractContract");
  }
};

// src/utils.ts
import { getAllDecodedLogs, getDecodedLogs } from "@fuel-ts/account";
import { ErrorCode, FuelError } from "@fuel-ts/errors";
function assert(condition, message) {
  if (!condition) {
    throw new FuelError(ErrorCode.TRANSACTION_ERROR, message);
  }
}
__name(assert, "assert");
function getAbisFromAllCalls(functionScopes) {
  return functionScopes.reduce((acc, funcScope, i) => {
    const { program, externalAbis } = funcScope.getCallConfig();
    if (i === 0) {
      acc.main = program.interface.jsonAbi;
      acc.otherContractsAbis = {};
    } else {
      acc.otherContractsAbis[program.id.toB256()] = program.interface.jsonAbi;
    }
    acc.otherContractsAbis = { ...acc.otherContractsAbis, ...externalAbis };
    return acc;
  }, {});
}
__name(getAbisFromAllCalls, "getAbisFromAllCalls");
var getResultLogs = /* @__PURE__ */ __name((receipts, mainCallConfig, functionScopes) => {
  if (!mainCallConfig) {
    return [];
  }
  const { main, otherContractsAbis } = getAbisFromAllCalls(functionScopes);
  return getDecodedLogs(receipts, main, otherContractsAbis);
}, "getResultLogs");
var getAllResultLogs = /* @__PURE__ */ __name((opts) => {
  const { receipts, mainCallConfig, functionScopes } = opts;
  if (!mainCallConfig) {
    return { logs: [], groupedLogs: {} };
  }
  const { main, otherContractsAbis } = getAbisFromAllCalls(functionScopes);
  return getAllDecodedLogs({ receipts, mainAbi: main, externalAbis: otherContractsAbis });
}, "getAllResultLogs");

// src/functions/invocation-scope.ts
import { coinQuantityfy } from "@fuel-ts/account";
import { ErrorCode as ErrorCode5, FuelError as FuelError5 } from "@fuel-ts/errors";

// src/functions/base-invocation-scope.ts
import {
  mergeQuantities,
  ScriptTransactionRequest,
  Wallet,
  setAndValidateGasAndFeeForAssembledTx
} from "@fuel-ts/account";
import { Address } from "@fuel-ts/address";
import { ErrorCode as ErrorCode4, FuelError as FuelError4 } from "@fuel-ts/errors";
import { bn as bn2 } from "@fuel-ts/math";
import { InputType, OutputType, TransactionType } from "@fuel-ts/transactions";
import { isDefined } from "@fuel-ts/utils";
import * as asm2 from "@fuels/vm-asm";
import { clone } from "ramda";

// src/contract-call-script.ts
import {
  WORD_SIZE as WORD_SIZE2,
  B256Coder,
  ASSET_ID_LEN as ASSET_ID_LEN2,
  BigNumberCoder,
  CONTRACT_ID_LEN as CONTRACT_ID_LEN2
} from "@fuel-ts/abi-coder";
import { ZeroBytes32 } from "@fuel-ts/address/configs";
import { ErrorCode as ErrorCode3, FuelError as FuelError3 } from "@fuel-ts/errors";
import { bn, toNumber } from "@fuel-ts/math";
import { ReceiptType as ReceiptType2 } from "@fuel-ts/transactions";
import { concat as concat2, arrayify as arrayify2 } from "@fuel-ts/utils";
import * as asm from "@fuels/vm-asm";

// src/instruction-set.ts
import { concat, hexlify } from "@fuel-ts/utils";
var InstructionSet = class {
  static {
    __name(this, "InstructionSet");
  }
  #operations;
  constructor(...args) {
    this.#operations = args || [];
  }
  entries() {
    return this.#operations;
  }
  push(...args) {
    this.#operations.push(...args);
  }
  concat(ops) {
    return this.#operations.concat(ops);
  }
  extend(ops) {
    this.#operations.push(...ops);
  }
  toBytes() {
    return concat(
      this.#operations.reduce((instructions, line) => {
        instructions.push(line.to_bytes());
        return instructions;
      }, [])
    );
  }
  toHex() {
    return hexlify(this.toBytes());
  }
  toString() {
    return `Program:
${JSON.stringify(this.#operations, null, 2)}`;
  }
  byteLength() {
    return this.toBytes().byteLength;
  }
};

// src/script-request.ts
import {
  ASSET_ID_LEN,
  CONTRACT_ID_LEN,
  SCRIPT_FIXED_SIZE,
  WORD_SIZE,
  calculateVmTxMemory
} from "@fuel-ts/abi-coder";
import { extractTxError } from "@fuel-ts/account";
import { ErrorCode as ErrorCode2, FuelError as FuelError2 } from "@fuel-ts/errors";
import { ReceiptType } from "@fuel-ts/transactions";
import { arrayify } from "@fuel-ts/utils";
var calculateScriptDataBaseOffset = /* @__PURE__ */ __name((maxInputs) => SCRIPT_FIXED_SIZE + calculateVmTxMemory({ maxInputs }), "calculateScriptDataBaseOffset");
var POINTER_DATA_OFFSET = WORD_SIZE + ASSET_ID_LEN + CONTRACT_ID_LEN + WORD_SIZE + WORD_SIZE;
function callResultToScriptResult(callResult) {
  const receipts = [...callResult.receipts];
  let scriptResultReceipt;
  let returnReceipt;
  receipts.forEach((receipt) => {
    if (receipt.type === ReceiptType.ScriptResult) {
      scriptResultReceipt = receipt;
    } else if (receipt.type === ReceiptType.Return || receipt.type === ReceiptType.ReturnData || receipt.type === ReceiptType.Revert) {
      returnReceipt = receipt;
    }
  });
  if (!scriptResultReceipt || !returnReceipt) {
    throw new FuelError2(ErrorCode2.SCRIPT_REVERTED, `Transaction reverted.`);
  }
  const scriptResult = {
    code: scriptResultReceipt.result,
    gasUsed: scriptResultReceipt.gasUsed,
    receipts,
    scriptResultReceipt,
    returnReceipt,
    callResult
  };
  return scriptResult;
}
__name(callResultToScriptResult, "callResultToScriptResult");
function decodeCallResult(callResult, decoder, logs = [], groupedLogs = {}) {
  try {
    const scriptResult = callResultToScriptResult(callResult);
    return decoder(scriptResult);
  } catch (error) {
    if (error.code === ErrorCode2.SCRIPT_REVERTED) {
      const statusReason = callResult?.dryRunStatus?.reason;
      throw extractTxError({
        logs,
        groupedLogs,
        receipts: callResult.receipts,
        statusReason
      });
    }
    throw error;
  }
}
__name(decodeCallResult, "decodeCallResult");
function callResultToInvocationResult(callResult, call2, logs, groupedLogs) {
  return decodeCallResult(
    callResult,
    (scriptResult) => {
      if (scriptResult.returnReceipt.type === ReceiptType.Revert) {
        throw new FuelError2(
          ErrorCode2.SCRIPT_REVERTED,
          `Script Reverted. Logs: ${JSON.stringify(logs)}`
        );
      }
      if (scriptResult.returnReceipt.type !== ReceiptType.Return && scriptResult.returnReceipt.type !== ReceiptType.ReturnData) {
        const { type } = scriptResult.returnReceipt;
        throw new FuelError2(
          ErrorCode2.SCRIPT_REVERTED,
          `Script Return Type [${type}] Invalid. Logs: ${JSON.stringify({
            logs,
            groupedLogs,
            receipt: scriptResult.returnReceipt
          })}`
        );
      }
      let value;
      if (scriptResult.returnReceipt.type === ReceiptType.Return) {
        value = scriptResult.returnReceipt.val;
      }
      if (scriptResult.returnReceipt.type === ReceiptType.ReturnData) {
        const decoded = call2.func.decodeOutput(scriptResult.returnReceipt.data);
        value = decoded[0];
      }
      return value;
    },
    logs
  );
}
__name(callResultToInvocationResult, "callResultToInvocationResult");
var ScriptRequest = class _ScriptRequest {
  static {
    __name(this, "ScriptRequest");
  }
  /**
   * The bytes of the script.
   */
  bytes;
  /**
   * A function to encode the script data.
   */
  scriptDataEncoder;
  /**
   * A function to decode the script result.
   */
  scriptResultDecoder;
  /**
   * Creates an instance of the ScriptRequest class.
   *
   * @param bytes - The bytes of the script.
   * @param scriptDataEncoder - The script data encoder function.
   * @param scriptResultDecoder - The script result decoder function.
   */
  constructor(bytes, scriptDataEncoder, scriptResultDecoder2) {
    this.bytes = arrayify(bytes);
    this.scriptDataEncoder = scriptDataEncoder;
    this.scriptResultDecoder = scriptResultDecoder2;
  }
  /**
   * Gets the script data offset for the given bytes.
   *
   * @param byteLength - The byte length of the script.
   * @param maxInputs - The maxInputs value from the chain's consensus params.
   * @returns The script data offset.
   */
  static getScriptDataOffsetWithScriptBytes(byteLength, maxInputs) {
    const scriptDataBaseOffset = calculateVmTxMemory({ maxInputs }) + SCRIPT_FIXED_SIZE;
    return scriptDataBaseOffset + byteLength;
  }
  /**
   * Gets the script data offset.
   *
   * @param maxInputs - The maxInputs value from the chain's consensus params.
   * @returns The script data offset.
   */
  getScriptDataOffset(maxInputs) {
    return _ScriptRequest.getScriptDataOffsetWithScriptBytes(this.bytes.length, maxInputs);
  }
  /**
   * Encodes the data for a script call.
   *
   * @param data - The script data.
   * @returns The encoded data.
   */
  encodeScriptData(data) {
    const callScript = this.scriptDataEncoder(data);
    if (ArrayBuffer.isView(callScript)) {
      return callScript;
    }
    this.bytes = arrayify(callScript.script);
    return callScript.data;
  }
  /**
   * Decodes the result of a script call.
   *
   * @param callResult - The CallResult from the script call.
   * @param logs - Optional logs associated with the decoding.
   * @returns The decoded result.
   */
  decodeCallResult(callResult, logs = []) {
    return decodeCallResult(callResult, this.scriptResultDecoder, logs);
  }
};

// src/contract-call-script.ts
var DEFAULT_OPCODE_PARAMS = {
  assetIdOffset: 0,
  amountOffset: 0,
  gasForwardedOffset: 0,
  callDataOffset: 0
};
var SCRIPT_WRAPPER_CONTRACT_ID = ZeroBytes32;
var getSingleCallInstructions = /* @__PURE__ */ __name(({
  callDataOffset,
  gasForwardedOffset,
  amountOffset,
  assetIdOffset
}) => {
  const inst = new InstructionSet(
    asm.movi(16, callDataOffset),
    asm.movi(17, amountOffset),
    asm.lw(17, 17, 0),
    asm.movi(18, assetIdOffset)
  );
  if (gasForwardedOffset) {
    inst.push(
      asm.movi(19, gasForwardedOffset),
      asm.lw(19, 19, 0),
      asm.call(16, 17, 18, 19)
    );
  } else {
    inst.push(asm.call(16, 17, 18, asm.RegId.cgas().to_u8()));
  }
  return inst;
}, "getSingleCallInstructions");
function getInstructions(offsets) {
  if (!offsets.length) {
    return new Uint8Array();
  }
  const multiCallInstructions = new InstructionSet();
  for (let i = 0; i < offsets.length; i += 1) {
    multiCallInstructions.extend(getSingleCallInstructions(offsets[i]).entries());
  }
  multiCallInstructions.push(asm.ret(1));
  return multiCallInstructions.toBytes();
}
__name(getInstructions, "getInstructions");
var isReturnType = /* @__PURE__ */ __name((type) => type === ReceiptType2.Return || type === ReceiptType2.ReturnData, "isReturnType");
var getMainCallReceipt = /* @__PURE__ */ __name((receipts, contractId) => receipts.find(
  ({ type, id, to }) => type === ReceiptType2.Call && id === SCRIPT_WRAPPER_CONTRACT_ID && to === contractId
), "getMainCallReceipt");
var scriptResultDecoder = /* @__PURE__ */ __name((contractId) => (result) => {
  if (toNumber(result.code) !== 0) {
    throw new FuelError3(ErrorCode3.SCRIPT_REVERTED, `Transaction reverted.`);
  }
  const mainCallResult = getMainCallReceipt(
    result.receipts,
    contractId.toB256()
  );
  const mainCallInstructionStart = bn(mainCallResult?.is);
  const receipts = result.receipts;
  return receipts.filter(({ type }) => isReturnType(type)).flatMap((receipt) => {
    if (!mainCallInstructionStart.eq(bn(receipt.is))) {
      return [];
    }
    if (receipt.type === ReceiptType2.Return) {
      return [new BigNumberCoder("u64").encode(receipt.val)];
    }
    if (receipt.type === ReceiptType2.ReturnData) {
      const encodedScriptReturn = arrayify2(receipt.data);
      return [encodedScriptReturn];
    }
    return [new Uint8Array()];
  });
}, "scriptResultDecoder");
var decodeContractCallScriptResult = /* @__PURE__ */ __name((callResult, contractId, logs = [], groupedLogs = {}) => decodeCallResult(callResult, scriptResultDecoder(contractId), logs, groupedLogs), "decodeContractCallScriptResult");
var getCallInstructionsLength = /* @__PURE__ */ __name((contractCalls) => contractCalls.reduce(
  (sum, call2) => {
    const offset = { ...DEFAULT_OPCODE_PARAMS };
    if (call2.gas) {
      offset.gasForwardedOffset = 1;
    }
    return sum + getSingleCallInstructions(offset).byteLength();
  },
  asm.Instruction.size()
  // placeholder for single RET instruction which is added later
), "getCallInstructionsLength");
var getContractCallScript = /* @__PURE__ */ __name((functionScopes, maxInputs) => new ScriptRequest(
  // Script to call the contract, start with stub size matching length of calls
  getInstructions(new Array(functionScopes.length).fill(DEFAULT_OPCODE_PARAMS)),
  (contractCalls) => {
    const TOTAL_CALLS = contractCalls.length;
    if (TOTAL_CALLS === 0) {
      return { data: new Uint8Array(), script: new Uint8Array() };
    }
    const callInstructionsLength = getCallInstructionsLength(contractCalls);
    const paddingLength = (8 - callInstructionsLength % 8) % 8;
    const paddedInstructionsLength = callInstructionsLength + paddingLength;
    const dataOffset = calculateScriptDataBaseOffset(maxInputs.toNumber()) + paddedInstructionsLength;
    const paramOffsets = [];
    let segmentOffset = dataOffset;
    const scriptData = [];
    for (let i = 0; i < TOTAL_CALLS; i += 1) {
      const call2 = contractCalls[i];
      const amountOffset = segmentOffset;
      const assetIdOffset = amountOffset + WORD_SIZE2;
      const callDataOffset = assetIdOffset + ASSET_ID_LEN2;
      const encodedSelectorOffset = callDataOffset + CONTRACT_ID_LEN2 + WORD_SIZE2 + WORD_SIZE2;
      const encodedArgsOffset = encodedSelectorOffset + call2.fnSelectorBytes.byteLength;
      const encodedArgs = arrayify2(call2.data);
      let gasForwardedOffset = 0;
      scriptData.push(new BigNumberCoder("u64").encode(call2.amount || 0));
      scriptData.push(new B256Coder().encode(call2.assetId?.toString() || ZeroBytes32));
      scriptData.push(call2.contractId.toBytes());
      scriptData.push(new BigNumberCoder("u64").encode(encodedSelectorOffset));
      scriptData.push(new BigNumberCoder("u64").encode(encodedArgsOffset));
      scriptData.push(call2.fnSelectorBytes);
      scriptData.push(encodedArgs);
      if (call2.gas) {
        scriptData.push(new BigNumberCoder("u64").encode(call2.gas));
        gasForwardedOffset = encodedArgsOffset + encodedArgs.byteLength;
      }
      const callParamOffsets = {
        amountOffset,
        assetIdOffset,
        gasForwardedOffset,
        callDataOffset
      };
      paramOffsets.push(callParamOffsets);
      segmentOffset = dataOffset + concat2(scriptData).byteLength;
    }
    const script = getInstructions(paramOffsets);
    const finalScriptData = concat2(scriptData);
    return { data: finalScriptData, script };
  },
  () => [new Uint8Array()]
), "getContractCallScript");

// src/response.ts
import { getGasUsedFromReceipts } from "@fuel-ts/account";
var extractInvocationResult = /* @__PURE__ */ __name((functionScopes, receipts, isMultiCall, logs, groupedLogs = {}) => {
  const mainCallConfig = functionScopes[0]?.getCallConfig();
  if (functionScopes.length === 1 && mainCallConfig && "bytes" in mainCallConfig.program) {
    return callResultToInvocationResult({ receipts }, mainCallConfig, logs, groupedLogs);
  }
  const encodedResults = decodeContractCallScriptResult(
    { receipts },
    (mainCallConfig?.program).id,
    logs,
    groupedLogs
  );
  const decodedResults = encodedResults.map((encodedResult, i) => {
    const { func } = functionScopes[i].getCallConfig();
    return func.decodeOutput(encodedResult)?.[0];
  });
  return isMultiCall ? decodedResults : decodedResults?.[0];
}, "extractInvocationResult");
var buildFunctionResult = /* @__PURE__ */ __name(async (params) => {
  const { funcScope, isMultiCall, program, transactionResponse } = params;
  const txResult = await transactionResponse.waitForResult();
  const { receipts } = txResult;
  const functionScopes = Array.isArray(funcScope) ? funcScope : [funcScope];
  const mainCallConfig = functionScopes[0]?.getCallConfig();
  const { logs, groupedLogs } = getAllResultLogs({ receipts, mainCallConfig, functionScopes });
  const value = extractInvocationResult(
    functionScopes,
    receipts,
    isMultiCall,
    logs,
    groupedLogs
  );
  const gasUsed = getGasUsedFromReceipts(receipts);
  const submitResult = {
    isMultiCall,
    functionScopes,
    value,
    program,
    transactionResult: txResult,
    transactionResponse,
    transactionId: transactionResponse.id,
    logs,
    groupedLogs,
    gasUsed
  };
  return submitResult;
}, "buildFunctionResult");
var buildPreConfirmationFunctionResult = /* @__PURE__ */ __name(async (params) => {
  const { funcScope, isMultiCall, program, transactionResponse } = params;
  const transactionResult = await transactionResponse.waitForPreConfirmation();
  const { receipts } = transactionResult;
  const functionScopes = Array.isArray(funcScope) ? funcScope : [funcScope];
  const mainCallConfig = functionScopes[0]?.getCallConfig();
  let logs;
  let groupedLogs;
  let gasUsed;
  let value;
  if (receipts) {
    ({ logs, groupedLogs } = getAllResultLogs({ receipts, mainCallConfig, functionScopes }));
    value = extractInvocationResult(functionScopes, receipts, isMultiCall, logs, groupedLogs);
    gasUsed = getGasUsedFromReceipts(receipts);
  }
  const submitResult = {
    isMultiCall,
    functionScopes,
    program,
    transactionResult,
    transactionResponse,
    transactionId: transactionResponse.id,
    logs,
    groupedLogs,
    gasUsed,
    value
  };
  return submitResult;
}, "buildPreConfirmationFunctionResult");
var buildDryRunResult = /* @__PURE__ */ __name((params) => {
  const { funcScopes, callResult, isMultiCall } = params;
  const { receipts } = callResult;
  const functionScopes = Array.isArray(funcScopes) ? funcScopes : [funcScopes];
  const mainCallConfig = functionScopes[0]?.getCallConfig();
  const { logs, groupedLogs } = getAllResultLogs({ receipts, mainCallConfig, functionScopes });
  const value = extractInvocationResult(
    functionScopes,
    receipts,
    isMultiCall,
    logs,
    groupedLogs
  );
  const gasUsed = getGasUsedFromReceipts(receipts);
  const submitResult = {
    functionScopes,
    callResult,
    isMultiCall,
    gasUsed,
    value
  };
  return submitResult;
}, "buildDryRunResult");

// src/functions/base-invocation-scope.ts
function createContractCall(funcScope) {
  const { program, args, forward, func, callParameters, externalAbis } = funcScope.getCallConfig();
  const data = func.encodeArguments(args);
  return {
    contractId: program.id,
    fnSelectorBytes: func.selectorBytes,
    data,
    assetId: forward?.assetId,
    amount: forward?.amount,
    gas: callParameters?.gasLimit,
    externalContractsAbis: externalAbis
  };
}
__name(createContractCall, "createContractCall");
var BaseInvocationScope = class {
  static {
    __name(this, "BaseInvocationScope");
  }
  transactionRequest;
  program;
  functionInvocationScopes = [];
  txParameters;
  requiredCoins = [];
  isMultiCall = false;
  hasCallParamsGasLimit = false;
  // flag to check if any of the callParams has gasLimit set
  externalAbis = {};
  /**
   * @deprecated - Should be removed with `addSigners`
   */
  addSignersCallback;
  /**
   * Constructs an instance of BaseInvocationScope.
   *
   * @param program - The abstract program to be invoked.
   * @param isMultiCall - A flag indicating whether the invocation is a multi-call.
   */
  constructor(program, isMultiCall) {
    this.program = program;
    this.isMultiCall = isMultiCall;
    this.transactionRequest = new ScriptTransactionRequest();
  }
  /**
   * Getter for the contract calls.
   *
   * @returns An array of contract calls.
   */
  get calls() {
    return this.functionInvocationScopes.map((funcScope) => createContractCall(funcScope));
  }
  /**
   * Updates the script request with the current contract calls.
   */
  async updateScriptRequest() {
    const provider = this.getProvider();
    const {
      consensusParameters: {
        txParameters: { maxInputs }
      }
    } = await provider.getChain();
    const contractCallScript = getContractCallScript(this.functionInvocationScopes, maxInputs);
    this.transactionRequest.setScript(contractCallScript, this.calls);
  }
  /**
   * Updates the transaction request with the current input/output.
   */
  updateContractInputAndOutput() {
    const calls = this.calls;
    calls.forEach((c) => {
      if (c.contractId) {
        this.transactionRequest.addContractInputAndOutput(c.contractId);
      }
      if (c.externalContractsAbis) {
        Object.keys(c.externalContractsAbis).forEach(
          (contractId) => this.transactionRequest.addContractInputAndOutput(new Address(contractId))
        );
      }
    });
  }
  /**
   * Gets the required coins for the transaction.
   *
   * @returns An array of required coin quantities.
   */
  getRequiredCoins() {
    const forwardingAssets = this.calls.map((call2) => ({
      assetId: String(call2.assetId),
      amount: bn2(call2.amount || 0)
    })).filter(({ assetId, amount }) => assetId && !bn2(amount).isZero());
    return forwardingAssets;
  }
  /**
   * Updates the required coins for the transaction.
   */
  updateRequiredCoins() {
    const assets = this.getRequiredCoins();
    const reduceForwardCoins = /* @__PURE__ */ __name((requiredCoins, { assetId, amount }) => {
      const currentAmount = requiredCoins.get(assetId)?.amount || bn2(0);
      return requiredCoins.set(assetId, {
        assetId: String(assetId),
        amount: currentAmount.add(amount)
      });
    }, "reduceForwardCoins");
    this.requiredCoins = Array.from(
      assets.reduce(reduceForwardCoins, /* @__PURE__ */ new Map()).values()
    );
  }
  /**
   * Adds a single call to the invocation scope.
   *
   * @param funcScope - The function scope to add.
   * @returns The current instance of the class.
   */
  addCall(funcScope) {
    this.addCalls([funcScope]);
    return this;
  }
  /**
   * Adds multiple calls to the invocation scope.
   *
   * @param funcScopes - An array of function scopes to add.
   * @returns The current instance of the class.
   */
  addCalls(funcScopes) {
    this.functionInvocationScopes.push(...funcScopes);
    this.updateContractInputAndOutput();
    this.updateRequiredCoins();
    return this;
  }
  /**
   * Prepares the transaction by updating the script request, required coins, and checking the gas limit.
   */
  async prepareTransaction() {
    await asm2.initWasm();
    await this.updateScriptRequest();
    this.updateRequiredCoins();
    this.checkGasLimitTotal();
    if (this.transactionRequest.type === TransactionType.Script) {
      this.transactionRequest.abis = getAbisFromAllCalls(this.functionInvocationScopes);
    }
  }
  /**
   * Checks if the total gas limit is within the acceptable range.
   */
  checkGasLimitTotal() {
    const gasLimitOnCalls = this.calls.reduce((total, call2) => total.add(call2.gas || 0), bn2(0));
    if (this.transactionRequest.gasLimit.eq(0)) {
      this.transactionRequest.gasLimit = gasLimitOnCalls;
    } else if (gasLimitOnCalls.gt(this.transactionRequest.gasLimit)) {
      throw new FuelError4(
        ErrorCode4.TRANSACTION_ERROR,
        "Transaction's gasLimit must be equal to or greater than the combined forwarded gas of all calls."
      );
    }
  }
  /**
   * Gets the transaction cost for dry running the transaction.
   *
   * @returns The transaction cost details.
   *
   * @deprecated Use contract.fundWithRequiredCoins instead
   * Check the migration guide https://docs.fuel.network/guide/assembling-transactions/migration-guide.html for more information.
   */
  async getTransactionCost() {
    const request = clone(await this.getTransactionRequest());
    const account = this.program.account ?? Wallet.generate({ provider: this.getProvider() });
    return account.getTransactionCost(request, {
      quantities: this.getRequiredCoins(),
      signatureCallback: this.addSignersCallback
    });
  }
  async fundWithRequiredCoins() {
    let request = await this.getTransactionRequest();
    request = clone(request);
    request.maxFee = bn2(0);
    request.gasLimit = bn2(0);
    const provider = this.getProvider();
    const account = this.program.account ?? Wallet.generate({ provider });
    const baseAssetId = await provider.getBaseAssetId();
    const outputQuantities = request.outputs.filter((o) => o.type === OutputType.Coin).map(({ amount, assetId }) => ({ assetId: String(assetId), amount: bn2(amount) }));
    const accountCoinQuantities = mergeQuantities(outputQuantities, this.requiredCoins);
    if (!accountCoinQuantities.length) {
      accountCoinQuantities.push({ assetId: baseAssetId, amount: bn2(0) });
    }
    let { assembledRequest, gasPrice } = await provider.assembleTx({
      request,
      feePayerAccount: account,
      accountCoinQuantities
    });
    assembledRequest = assembledRequest;
    await setAndValidateGasAndFeeForAssembledTx({
      gasPrice,
      provider,
      transactionRequest: assembledRequest,
      setGasLimit: this.txParameters?.gasLimit,
      setMaxFee: this.txParameters?.maxFee
    });
    return assembledRequest;
  }
  /**
   * @deprecated - Should be removed with `addSigners`
   */
  async legacyFundWithRequiredCoins() {
    let transactionRequest = await this.getTransactionRequest();
    transactionRequest = clone(transactionRequest);
    const txCost = await this.getTransactionCost();
    const { gasUsed, missingContractIds, outputVariables, maxFee } = txCost;
    this.setDefaultTxParams(transactionRequest, gasUsed, maxFee);
    missingContractIds.forEach((contractId) => {
      transactionRequest.addContractInputAndOutput(new Address(contractId));
    });
    transactionRequest.addVariableOutputs(outputVariables);
    await this.program.account?.fund(transactionRequest, txCost);
    if (this.addSignersCallback) {
      await this.addSignersCallback(transactionRequest);
    }
    return transactionRequest;
  }
  /**
   * Sets the transaction parameters.
   *
   * @param txParams - The transaction parameters to set.
   * @returns The current instance of the class.
   */
  txParams(txParams) {
    this.txParameters = txParams;
    const request = this.transactionRequest;
    request.tip = bn2(txParams.tip || request.tip);
    request.gasLimit = bn2(txParams.gasLimit || request.gasLimit);
    request.maxFee = txParams.maxFee ? bn2(txParams.maxFee) : request.maxFee;
    request.witnessLimit = txParams.witnessLimit ? bn2(txParams.witnessLimit) : request.witnessLimit;
    request.maturity = txParams.maturity || request.maturity;
    request.expiration = txParams.expiration || request.expiration;
    request.addVariableOutputs(this.txParameters?.variableOutputs || 0);
    return this;
  }
  /**
   * Adds contracts to the invocation scope.
   *
   * @param contracts - An array of contracts to add.
   * @returns The current instance of the class.
   */
  addContracts(contracts) {
    contracts.forEach((contract) => {
      if (typeof contract === "string") {
        this.transactionRequest.addContractInputAndOutput(new Address(contract));
      } else {
        this.transactionRequest.addContractInputAndOutput(contract.id);
        this.externalAbis[contract.id.toB256()] = contract.interface.jsonAbi;
      }
    });
    return this;
  }
  /**
   * Adds an asset transfer to an Account on the contract call transaction request.
   *
   * @param transferParams - The object representing the transfer to be made.
   * @returns The current instance of the class.
   */
  addTransfer(transferParams) {
    const { amount, destination, assetId } = transferParams;
    this.transactionRequest = this.transactionRequest.addCoinOutput(
      new Address(destination),
      amount,
      assetId
    );
    return this;
  }
  /**
   * Adds multiple transfers to the contract call transaction request.
   *
   * @param transferParams - An array of `TransferParams` objects representing the transfers to be made.
   * @returns The current instance of the class.
   */
  addBatchTransfer(transferParams) {
    transferParams.forEach(({ destination, amount, assetId }) => {
      this.transactionRequest = this.transactionRequest.addCoinOutput(
        new Address(destination),
        amount,
        assetId
      );
    });
    return this;
  }
  /**
   * Adds signers to the transaction request.
   *
   * @param signers - The signers to add.
   * @returns The current instance of the class.
   *
   * @deprecated This method is deprecated and will be removed in a future versions.
   * All signatures should be manually added to the transaction request witnesses. If your
   * Sway program relies on in-code signature validation, visit this guide:
   * https://docs.fuel.network/guides/cookbook/sway-script-with-signature-validation.html
   */
  addSigners(signers) {
    this.addSignersCallback = (transactionRequest) => transactionRequest.addAccountWitnesses(signers);
    return this;
  }
  /**
   * Prepares and returns the transaction request object.
   *
   * @returns The prepared transaction request.
   */
  async getTransactionRequest() {
    await this.prepareTransaction();
    return this.transactionRequest;
  }
  /**
   * Submits the contract call transaction and returns a promise that resolves to an object
   * containing the transaction ID and a function to wait for the result. The promise will resolve
   * as soon as the transaction is submitted to the node.
   *
   * @returns A promise that resolves to an object containing:
   * - `transactionId`: The ID of the submitted transaction.
   * - `waitForResult`: A function that waits for the transaction result.
   * @template T - The type of the return value.
   */
  async call() {
    assert(this.program.account, "Wallet is required!");
    let transactionRequest = await this.getTransactionRequest();
    if (this.addSignersCallback) {
      transactionRequest = await this.legacyFundWithRequiredCoins();
    } else {
      transactionRequest = await this.fundWithRequiredCoins();
    }
    const response = await this.program.account.sendTransaction(transactionRequest, {
      estimateTxDependencies: false
    });
    const transactionId = response.id;
    return {
      transactionId,
      waitForResult: /* @__PURE__ */ __name(async () => buildFunctionResult({
        funcScope: this.functionInvocationScopes,
        isMultiCall: this.isMultiCall,
        program: this.program,
        transactionResponse: response
      }), "waitForResult"),
      waitForPreConfirmation: /* @__PURE__ */ __name(async () => buildPreConfirmationFunctionResult({
        funcScope: this.functionInvocationScopes,
        isMultiCall: this.isMultiCall,
        program: this.program,
        transactionResponse: response
      }), "waitForPreConfirmation")
    };
  }
  /**
   * Simulates a transaction.
   *
   * @returns The result of the invocation call.
   */
  async simulate() {
    assert(this.program.account, "Wallet is required!");
    if (!("populateTransactionWitnessesSignature" in this.program.account)) {
      throw new FuelError4(
        ErrorCode4.ABI_MAIN_METHOD_MISSING,
        "An unlocked wallet is required to simulate a contract call."
      );
    }
    const transactionRequest = await this.fundWithRequiredCoins();
    const callResult = await this.program.account.simulateTransaction(transactionRequest, {
      estimateTxDependencies: false
    });
    return buildDryRunResult({
      funcScopes: this.functionInvocationScopes,
      callResult,
      isMultiCall: this.isMultiCall
    });
  }
  /**
   * Executes a transaction in dry run mode.
   *
   * @returns The result of the invocation call.
   *
   * @deprecated Use .get instead
   */
  async dryRun() {
    const { receipts } = await this.getTransactionCost();
    const callResult = {
      receipts
    };
    return buildDryRunResult({
      funcScopes: this.functionInvocationScopes,
      callResult,
      isMultiCall: this.isMultiCall
    });
  }
  async get() {
    let request = await this.getTransactionRequest();
    request = clone(request);
    request.maxFee = bn2(0);
    request.gasLimit = bn2(0);
    request.inputs = request.inputs.filter((i) => i.type !== InputType.Coin);
    const provider = this.getProvider();
    const account = this.program.account ?? Wallet.generate({ provider });
    const baseAssetId = await provider.getBaseAssetId();
    const allQuantities = request.outputs.filter((o) => o.type === OutputType.Coin).map(({ amount, assetId }) => ({ assetId: String(assetId), amount: bn2(amount) })).concat(this.requiredCoins);
    const resources = account.generateFakeResources(allQuantities);
    const utxoForBaseAssetId = resources.find((utxo) => utxo.assetId === baseAssetId);
    const amountForFee = bn2("1000000000000000");
    if (!utxoForBaseAssetId) {
      const [baseAssetResource] = account.generateFakeResources([
        { assetId: baseAssetId, amount: amountForFee }
      ]);
      resources.push(baseAssetResource);
    } else {
      utxoForBaseAssetId.amount = utxoForBaseAssetId.amount.add(amountForFee);
    }
    request.addResources(resources);
    const { receipts } = await provider.assembleTx({
      request,
      feePayerAccount: account
    });
    return buildDryRunResult({
      funcScopes: this.functionInvocationScopes,
      callResult: { receipts },
      isMultiCall: this.isMultiCall
    });
  }
  getProvider() {
    const provider = this.program.provider;
    return provider;
  }
  /**
   * Obtains the ID of a transaction.
   *
   * @param chainId - the chainId to use to hash the transaction with
   * @returns the ID of the transaction.
   */
  async getTransactionId(chainId) {
    const chainIdToHash = chainId ?? await this.getProvider().getChainId();
    const transactionRequest = await this.getTransactionRequest();
    return transactionRequest.getTransactionId(chainIdToHash);
  }
  /**
   * In case the gasLimit is *not* set by the user, this method sets a default value.
   */
  setDefaultTxParams(transactionRequest, gasUsed, maxFee) {
    const gasLimitSpecified = isDefined(this.txParameters?.gasLimit) || this.hasCallParamsGasLimit;
    const maxFeeSpecified = isDefined(this.txParameters?.maxFee);
    const { gasLimit: setGasLimit, maxFee: setMaxFee } = transactionRequest;
    if (!gasLimitSpecified) {
      transactionRequest.gasLimit = gasUsed;
    } else if (setGasLimit.lt(gasUsed)) {
      throw new FuelError4(
        ErrorCode4.GAS_LIMIT_TOO_LOW,
        `Gas limit '${setGasLimit}' is lower than the required: '${gasUsed}'.`
      );
    }
    if (!maxFeeSpecified) {
      transactionRequest.maxFee = maxFee;
    } else if (maxFee.gt(setMaxFee)) {
      throw new FuelError4(
        ErrorCode4.MAX_FEE_TOO_LOW,
        `Max fee '${setMaxFee}' is lower than the required: '${maxFee}'.`
      );
    }
  }
};

// src/functions/invocation-scope.ts
var FunctionInvocationScope = class extends BaseInvocationScope {
  static {
    __name(this, "FunctionInvocationScope");
  }
  func;
  callParameters;
  forward;
  args;
  /**
   * Constructs an instance of FunctionInvocationScope.
   *
   * @param program - The program.
   * @param func - The function fragment.
   * @param args - The arguments.
   */
  constructor(program, func, args) {
    super(program, false);
    this.func = func;
    this.args = args || [];
    this.setArguments(...args);
    super.addCall(this);
  }
  /**
   * Gets the call configuration.
   *
   * @returns The call configuration.
   */
  getCallConfig() {
    return {
      func: this.func,
      program: this.program,
      callParameters: this.callParameters,
      txParameters: this.txParameters,
      forward: this.forward,
      args: this.args,
      externalAbis: this.externalAbis
    };
  }
  /**
   * Sets the arguments for the function invocation.
   *
   * @param args - The arguments.
   * @returns The instance of FunctionInvocationScope.
   */
  setArguments(...args) {
    this.args = args || [];
    return this;
  }
  /**
   * Sets the call parameters for the function invocation.
   *
   * @param callParams - The call parameters.
   * @returns The instance of FunctionInvocationScope.
   * @throws If the function is not payable and forward is set.
   */
  callParams(callParams) {
    if (!this.hasCallParamsGasLimit && callParams?.gasLimit !== void 0) {
      this.hasCallParamsGasLimit = true;
    }
    this.callParameters = callParams;
    if (callParams?.forward) {
      if (!this.func.attributes.find((attr) => attr.name === "payable")) {
        throw new FuelError5(
          ErrorCode5.TRANSACTION_ERROR,
          `The target function ${this.func.name} cannot accept forwarded funds as it's not marked as 'payable'.`
        );
      }
      this.forward = coinQuantityfy(callParams.forward);
    }
    this.setArguments(...this.args);
    this.updateRequiredCoins();
    return this;
  }
};

// src/functions/multicall-scope.ts
var MultiCallInvocationScope = class extends BaseInvocationScope {
  static {
    __name(this, "MultiCallInvocationScope");
  }
  /**
   * Constructs an instance of MultiCallInvocationScope.
   *
   * @param contract - The contract.
   * @param funcScopes - An array of function invocation scopes.
   */
  constructor(contract, funcScopes) {
    super(contract, true);
    this.addCalls(funcScopes);
  }
  /**
   * Adds a single function invocation scope to the multi-call invocation scope.
   *
   * @param funcScope - The function invocation scope.
   * @returns The instance of MultiCallInvocationScope.
   */
  addCall(funcScope) {
    return super.addCalls([funcScope]);
  }
  /**
   * Adds multiple function invocation scopes to the multi-call invocation scope.
   *
   * @param funcScopes - An array of function invocation scopes.
   * @returns The instance of MultiCallInvocationScope.
   */
  addCalls(funcScopes) {
    return super.addCalls(funcScopes);
  }
};

// src/contract.ts
import { Interface } from "@fuel-ts/abi-coder";
import { Address as Address2 } from "@fuel-ts/address";
var Contract = class {
  static {
    __name(this, "Contract");
  }
  /**
   * The unique contract identifier.
   */
  id;
  /**
   * The provider for interacting with the contract.
   */
  provider;
  /**
   * The contract's ABI interface.
   */
  interface;
  /**
   * The account associated with the contract, if available.
   */
  account;
  /**
   * A collection of functions available on the contract.
   */
  functions = {};
  /**
   * Creates an instance of the Contract class.
   *
   * @param id - The contract's address.
   * @param abi - The contract's ABI (JSON ABI or Interface instance).
   * @param accountOrProvider - The account or provider for interaction.
   */
  constructor(id, abi, accountOrProvider) {
    this.interface = abi instanceof Interface ? abi : new Interface(abi);
    this.id = new Address2(id);
    if (accountOrProvider && "provider" in accountOrProvider) {
      this.provider = accountOrProvider.provider;
      this.account = accountOrProvider;
    } else {
      this.provider = accountOrProvider;
      this.account = null;
    }
    Object.keys(this.interface.functions).forEach((name) => {
      const fragment = this.interface.getFunction(name);
      Object.defineProperty(this.functions, fragment.name, {
        value: this.buildFunction(fragment),
        writable: false
      });
    });
  }
  /**
   * Build a function invocation scope for the provided function fragment.
   *
   * @param func - The function fragment to build a scope for.
   * @returns A function that creates a FunctionInvocationScope.
   */
  buildFunction(func) {
    return (() => {
      const funcInvocationScopeCreator = /* @__PURE__ */ __name((...args) => new FunctionInvocationScope(this, func, args), "funcInvocationScopeCreator");
      Object.defineProperty(funcInvocationScopeCreator, "isReadOnly", {
        value: /* @__PURE__ */ __name(() => func.isReadOnly(), "value"),
        writable: false
      });
      return funcInvocationScopeCreator;
    })();
  }
  /**
   * Create a multi-call invocation scope for the provided function invocation scopes.
   *
   * @param calls - An array of FunctionInvocationScopes to execute in a batch.
   * @returns A MultiCallInvocationScope instance.
   */
  multiCall(calls) {
    return new MultiCallInvocationScope(this, calls);
  }
  /**
   * Get the balance for a given asset ID for this contract.
   *
   * @param assetId - The specified asset ID.
   * @returns The balance of the contract for the specified asset.
   */
  getBalance(assetId) {
    return this.provider.getContractBalance(this.id, assetId);
  }
};
export {
  AbstractContract,
  AbstractProgram,
  Contract,
  FunctionInvocationScope,
  InstructionSet,
  MultiCallInvocationScope,
  ScriptRequest,
  assert,
  buildDryRunResult,
  buildFunctionResult,
  buildPreConfirmationFunctionResult,
  extractInvocationResult,
  getAbisFromAllCalls,
  getAllResultLogs,
  getResultLogs
};
