// @arcana/ca-common/dist/cjs/proto/grpc.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.7.0
//   protoc               v6.31.0
// source: grpc.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcWebError = exports.GrpcWebImpl = exports.QueryPriceOracleDataDesc = exports.QueryProtocolFeesDesc = exports.QuerySettlementAllDesc = exports.QuerySettlementDesc = exports.QuerySolverDataAllDesc = exports.QuerySolverDataDesc = exports.QueryRequestForFundsByAddressDesc = exports.QueryRequestForFundsAllDesc = exports.QueryRequestForFundsDesc = exports.QueryDesc = exports.QueryClientImpl = exports.protobufPackage = void 0;
/* eslint-disable */
const grpc_web_1 = require("@improbable-eng/grpc-web");
const browser_headers_1 = require("browser-headers");
const definition_1 = require("./definition");
exports.protobufPackage = "xarchain.chainabstraction";
class QueryClientImpl {
    rpc;
    constructor(rpc) {
        this.rpc = rpc;
        this.RequestForFunds = this.RequestForFunds.bind(this);
        this.RequestForFundsAll = this.RequestForFundsAll.bind(this);
        this.RequestForFundsByAddress = this.RequestForFundsByAddress.bind(this);
        this.SolverData = this.SolverData.bind(this);
        this.SolverDataAll = this.SolverDataAll.bind(this);
        this.Settlement = this.Settlement.bind(this);
        this.SettlementAll = this.SettlementAll.bind(this);
        this.ProtocolFees = this.ProtocolFees.bind(this);
        this.PriceOracleData = this.PriceOracleData.bind(this);
    }
    RequestForFunds(request, metadata) {
        return this.rpc.unary(exports.QueryRequestForFundsDesc, definition_1.QueryGetRequestForFundsRequest.fromPartial(request), metadata);
    }
    RequestForFundsAll(request, metadata) {
        return this.rpc.unary(exports.QueryRequestForFundsAllDesc, definition_1.QueryAllRequestForFundsRequest.fromPartial(request), metadata);
    }
    RequestForFundsByAddress(request, metadata) {
        return this.rpc.unary(exports.QueryRequestForFundsByAddressDesc, definition_1.QueryRequestForFundsByAddressRequest.fromPartial(request), metadata);
    }
    SolverData(request, metadata) {
        return this.rpc.unary(exports.QuerySolverDataDesc, definition_1.QueryGetSolverDataRequest.fromPartial(request), metadata);
    }
    SolverDataAll(request, metadata) {
        return this.rpc.unary(exports.QuerySolverDataAllDesc, definition_1.QueryAllSolverDataRequest.fromPartial(request), metadata);
    }
    Settlement(request, metadata) {
        return this.rpc.unary(exports.QuerySettlementDesc, definition_1.QueryGetSettlementRequest.fromPartial(request), metadata);
    }
    SettlementAll(request, metadata) {
        return this.rpc.unary(exports.QuerySettlementAllDesc, definition_1.QueryAllSettlementRequest.fromPartial(request), metadata);
    }
    ProtocolFees(request, metadata) {
        return this.rpc.unary(exports.QueryProtocolFeesDesc, definition_1.QueryGetProtocolFeesRequest.fromPartial(request), metadata);
    }
    PriceOracleData(request, metadata) {
        return this.rpc.unary(exports.QueryPriceOracleDataDesc, definition_1.QueryGetPriceOracleDataRequest.fromPartial(request), metadata);
    }
}
exports.QueryClientImpl = QueryClientImpl;
exports.QueryDesc = { serviceName: "xarchain.chainabstraction.Query" };
exports.QueryRequestForFundsDesc = {
    methodName: "RequestForFunds",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return definition_1.QueryGetRequestForFundsRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = definition_1.QueryGetRequestForFundsResponse.decode(data);
            return {
                ...value,
                toObject() {
                    return value;
                },
            };
        },
    },
};
exports.QueryRequestForFundsAllDesc = {
    methodName: "RequestForFundsAll",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return definition_1.QueryAllRequestForFundsRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = definition_1.QueryAllRequestForFundsResponse.decode(data);
            return {
                ...value,
                toObject() {
                    return value;
                },
            };
        },
    },
};
exports.QueryRequestForFundsByAddressDesc = {
    methodName: "RequestForFundsByAddress",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return definition_1.QueryRequestForFundsByAddressRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = definition_1.QueryAllRequestForFundsResponse.decode(data);
            return {
                ...value,
                toObject() {
                    return value;
                },
            };
        },
    },
};
exports.QuerySolverDataDesc = {
    methodName: "SolverData",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return definition_1.QueryGetSolverDataRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = definition_1.QueryGetSolverDataResponse.decode(data);
            return {
                ...value,
                toObject() {
                    return value;
                },
            };
        },
    },
};
exports.QuerySolverDataAllDesc = {
    methodName: "SolverDataAll",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return definition_1.QueryAllSolverDataRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = definition_1.QueryAllSolverDataResponse.decode(data);
            return {
                ...value,
                toObject() {
                    return value;
                },
            };
        },
    },
};
exports.QuerySettlementDesc = {
    methodName: "Settlement",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return definition_1.QueryGetSettlementRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = definition_1.QueryGetSettlementResponse.decode(data);
            return {
                ...value,
                toObject() {
                    return value;
                },
            };
        },
    },
};
exports.QuerySettlementAllDesc = {
    methodName: "SettlementAll",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return definition_1.QueryAllSettlementRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = definition_1.QueryAllSettlementResponse.decode(data);
            return {
                ...value,
                toObject() {
                    return value;
                },
            };
        },
    },
};
exports.QueryProtocolFeesDesc = {
    methodName: "ProtocolFees",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return definition_1.QueryGetProtocolFeesRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = definition_1.QueryGetProtocolFeesResponse.decode(data);
            return {
                ...value,
                toObject() {
                    return value;
                },
            };
        },
    },
};
exports.QueryPriceOracleDataDesc = {
    methodName: "PriceOracleData",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return definition_1.QueryGetPriceOracleDataRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = definition_1.QueryGetPriceOracleDataResponse.decode(data);
            return {
                ...value,
                toObject() {
                    return value;
                },
            };
        },
    },
};
class GrpcWebImpl {
    host;
    options;
    constructor(host, options) {
        this.host = host;
        this.options = options;
    }
    unary(methodDesc, _request, metadata) {
        const request = { ..._request, ...methodDesc.requestType };
        const maybeCombinedMetadata = metadata && this.options.metadata
            ? new browser_headers_1.BrowserHeaders({
                ...this.options?.metadata.headersMap,
                ...metadata?.headersMap,
            })
            : (metadata ?? this.options.metadata);
        return new Promise((resolve, reject) => {
            grpc_web_1.grpc.unary(methodDesc, {
                request,
                host: this.host,
                metadata: maybeCombinedMetadata ?? {},
                ...(this.options.transport !== undefined
                    ? { transport: this.options.transport }
                    : {}),
                debug: this.options.debug ?? false,
                onEnd: function (response) {
                    if (response.status === grpc_web_1.grpc.Code.OK) {
                        resolve(response.message.toObject());
                    }
                    else {
                        const err = new GrpcWebError(response.statusMessage, response.status, response.trailers);
                        reject(err);
                    }
                },
            });
        });
    }
}
exports.GrpcWebImpl = GrpcWebImpl;
class GrpcWebError extends globalThis.Error {
    code;
    metadata;
    constructor(message, code, metadata) {
        super(message);
        this.code = code;
        this.metadata = metadata;
    }
}
exports.GrpcWebError = GrpcWebError;
