// @arcana/ca-common/dist/cjs/vaultcontracts/vaultcontracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VaultContractMap = exports.Environment = void 0;
const viem_1 = require("viem");
const data_1 = require("../data");
const definition_1 = require("../proto/definition");
var Environment;
(function (Environment) {
    Environment[Environment["FOLLY"] = 0] = "FOLLY";
    Environment[Environment["CERISE"] = 1] = "CERISE";
    Environment[Environment["CORAL"] = 2] = "CORAL";
    Environment[Environment["JADE"] = 3] = "JADE";
})(Environment || (exports.Environment = Environment = {}));
const dataSets = new Map([
    [
        Environment.FOLLY,
        [
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 421614),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 11155420),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 80002),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 84532),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
        ],
    ],
    [
        Environment.CERISE,
        [
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 137),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 10),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 42161),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 8453),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 43114),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 534352),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 8217),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 50104),
                "0x0A8eb0237524D1A8Fa8cbecF49e54FE627Ed781f",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.FUEL, 9889),
                "0x6cd9b8d7e13762f4cb98cbd733640138aeb65395c4b38d84a7d46f6f6c6b42e4",
            ],
        ],
    ],
    [
        Environment.CORAL,
        [
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 1),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 10),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 137),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 42161),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 534352),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 8453),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 43114),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 999),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 8217),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.ETHEREUM, 50104),
                "0xB61fAdeBccCb15823b64bf47829d32eeb4A08930",
            ],
            [
                (0, data_1.encodeChainID36)(definition_1.Universe.FUEL, 9889),
                "0x18bf9116890a9ca862b3dd2157314e818b3ba1434b21039592a84ab23740a588",
            ],
        ],
    ],
]);
class VaultContractMap {
    map = new Map();
    constructor(environment) {
        const src = dataSets.get(environment);
        if (src == null) {
            throw new Error("Environment not found");
        }
        for (const tuple of src) {
            this.map.set((0, viem_1.toHex)(tuple[0]), (0, data_1.zeroExtendBufToGivenSize)((0, viem_1.hexToBytes)(tuple[1]), 32));
        }
    }
    getFromChainID36(key) {
        return this.map.get((0, viem_1.toHex)(key));
    }
    getFromOmniversalChainID(key) {
        return this.getFromChainID36(key.toBytes());
    }
    *entries() {
        for (const [key, value] of this.map.entries()) {
            yield [data_1.OmniversalChainID.fromChainID36((0, viem_1.hexToBytes)(key)), value];
        }
    }
}
exports.VaultContractMap = VaultContractMap;
