// @fuel-ts/versions/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  checkFuelCoreVersionCompatibility: () => checkFuelCoreVersionCompatibility,
  eq: () => eq,
  gt: () => gt,
  gte: () => gte,
  majorEq: () => majorEq,
  minorEq: () => minorEq,
  patchEq: () => patchEq,
  versions: () => versions
});
module.exports = __toCommonJS(src_exports);

// src/lib/getBuiltinVersions.ts
function getBuiltinVersions() {
  return {
    FUEL_CORE: "0.43.1",
    FORC: "0.68.1",
    FUELS: "0.100.6"
  };
}
__name(getBuiltinVersions, "getBuiltinVersions");

// src/lib/semver.ts
function parseVersion(version) {
  const [major, minor, patch] = version.split(".").map((v) => parseInt(v, 10));
  return { major, minor, patch };
}
__name(parseVersion, "parseVersion");
function versionDiffs(version1, version2) {
  const semver1 = parseVersion(version1);
  const semver2 = parseVersion(version2);
  const major = semver1.major - semver2.major;
  const minor = semver1.minor - semver2.minor;
  const patch = semver1.patch - semver2.patch;
  return {
    major,
    minor,
    patch,
    fullVersionDiff: major || minor || patch
  };
}
__name(versionDiffs, "versionDiffs");
function gt(version1, version2) {
  const { fullVersionDiff } = versionDiffs(version1, version2);
  return fullVersionDiff > 0;
}
__name(gt, "gt");
function eq(version1, version2) {
  const { fullVersionDiff } = versionDiffs(version1, version2);
  return fullVersionDiff === 0;
}
__name(eq, "eq");
function gte(version1, version2) {
  const { fullVersionDiff } = versionDiffs(version1, version2);
  return fullVersionDiff >= 0;
}
__name(gte, "gte");
function majorEq(version1, version2) {
  const { major } = versionDiffs(version1, version2);
  return major === 0;
}
__name(majorEq, "majorEq");
function minorEq(version1, version2) {
  const { minor } = versionDiffs(version1, version2);
  return minor === 0;
}
__name(minorEq, "minorEq");
function patchEq(version1, version2) {
  const { patch } = versionDiffs(version1, version2);
  return patch === 0;
}
__name(patchEq, "patchEq");

// src/lib/checkFuelCoreVersionCompatibility.ts
function checkFuelCoreVersionCompatibility(networkVersion) {
  const { FUEL_CORE: supportedVersion } = getBuiltinVersions();
  if (/^\d+\.\d+\.\d+\D+/m.test(networkVersion)) {
    console.warn(`You're running against an unreleased fuel-core version: ${networkVersion}. Things may work as expected, but it's not guaranteed. Please use a released version.      
This unreleased fuel-core build may include features and updates not yet supported by this version of the TS-SDK.`);
  }
  return {
    supportedVersion,
    isMajorSupported: majorEq(networkVersion, supportedVersion),
    isMinorSupported: minorEq(networkVersion, supportedVersion),
    isPatchSupported: patchEq(networkVersion, supportedVersion)
  };
}
__name(checkFuelCoreVersionCompatibility, "checkFuelCoreVersionCompatibility");

// src/index.ts
var versions = getBuiltinVersions();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  checkFuelCoreVersionCompatibility,
  eq,
  gt,
  gte,
  majorEq,
  minorEq,
  patchEq,
  versions
});
