// @arcana/ca-common/dist/cjs/rff/rff.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OmniversalRFF = void 0;
const viem_1 = require("viem");
const fuels_1 = require("fuels");
const data_1 = require("../data");
const fuel_1 = require("./fuel");
class OmniversalRFF {
    protobufRFF;
    evmRFF;
    fuelRFF;
    constructor(protobufRFF) {
        this.protobufRFF = protobufRFF;
    }
    asEVMRFF() {
        if (this.evmRFF == null) {
            this.evmRFF = {
                sources: this.protobufRFF.sources.map(s => ({
                    universe: s.universe,
                    chainID: (0, viem_1.bytesToBigInt)(s.chainID),
                    tokenAddress: (0, data_1.ezPadTo32Hex)(s.tokenAddress),
                    value: (0, viem_1.bytesToBigInt)(s.value),
                })),
                destinationUniverse: this.protobufRFF.destinationUniverse,
                destinationChainID: (0, viem_1.bytesToBigInt)(this.protobufRFF.destinationChainID),
                destinations: this.protobufRFF.destinations.map(d => ({
                    tokenAddress: (0, data_1.ezPadTo32Hex)(d.tokenAddress),
                    value: (0, viem_1.bytesToBigInt)(d.value),
                })),
                nonce: (0, viem_1.bytesToBigInt)(this.protobufRFF.nonce),
                expiry: BigInt(this.protobufRFF.expiry.toString()),
                parties: this.protobufRFF.signatureData.map(sd => ({
                    universe: sd.universe,
                    address_: (0, data_1.ezPadTo32Hex)(sd.address),
                }))
            };
        }
        return this.evmRFF;
    }
    asFuelRFF() {
        if (this.fuelRFF == null) {
            this.fuelRFF = {
                sources: this.protobufRFF.sources.map(s => ({
                    universe: (0, fuel_1.protobufUniverseToFuelUniverse)(s.universe),
                    chain_id: new fuels_1.BN(s.chainID),
                    asset_id: {
                        bits: (0, data_1.ezPadTo32Hex)(s.tokenAddress),
                    },
                    value: new fuels_1.BN(s.value),
                })),
                destination_chain_id: new fuels_1.BN(this.protobufRFF.destinationChainID),
                destination_universe: (0, fuel_1.protobufUniverseToFuelUniverse)(this.protobufRFF.destinationUniverse),
                destinations: this.protobufRFF.destinations.map(d => ({
                    asset_id: {
                        bits: (0, data_1.ezPadTo32Hex)(d.tokenAddress)
                    },
                    value: new fuels_1.BN(d.value)
                })),
                expiry: new fuels_1.BN(this.protobufRFF.expiry.toBytesBE()),
                nonce: new fuels_1.BN(this.protobufRFF.nonce),
                parties: this.protobufRFF.signatureData.map(sd => ({
                    universe: (0, fuel_1.protobufUniverseToFuelUniverse)(sd.universe),
                    address: {
                        bits: (0, data_1.ezPadTo32Hex)(sd.address),
                    }
                })),
            };
        }
        return this.fuelRFF;
    }
    asProtobufRFF() {
        return this.protobufRFF;
    }
}
exports.OmniversalRFF = OmniversalRFF;
