// @arcana/ca-common/dist/cjs/balances/ub-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBalances = getBalances;
const tslib_1 = require("tslib");
const decimal_js_1 = tslib_1.__importDefault(require("decimal.js"));
const definition_1 = require("../proto/definition");
const types_1 = require("../types");
const data_1 = require("../data");
const viem_1 = require("viem");
async function getBalances(vscURL, universe, address) {
    const resp = await (0, types_1.msgpackableAxios)({
        baseURL: vscURL,
        url: `/api/v1/get-balance/${definition_1.Universe[universe]}/${(0, viem_1.bytesToHex)(address)}`,
    });
    const { data } = resp;
    return data.balances.map(bal => ({
        chainID: new data_1.OmniversalChainID(bal.universe, bal.chain_id),
        totalValue: new decimal_js_1.default(bal.total_usd),
        currencywise: bal.currencies.map(cur => ({
            tokenAddress: cur.token_address,
            amount: new decimal_js_1.default(cur.balance),
            value: new decimal_js_1.default(cur.value)
        }))
    }));
}
